### DISCUSSION TITLE
Soft-Body questions

### DISCUSSION THREAD

**Discussion by jankrassnigg** [jrouwe/JoltPhysics#850] [Q&A] [answered]
While playing with soft bodies, I ran into a few issues and wanted to ask:

1) Is it possible (or planned) to give soft-bodies a "thickness" ? This would be useful for cloth, where you only have a single layer. Even if they perfectly collide with geometry without passing through, the rendering often results in z-fighting. It would be very practical, if a soft-body could have a thickness of a couple of centimeters, so that it stays slightly away from other geometry.

2) Is it possible to have the soft-bodies collide with other (dynamic) objects, but NOT push them? For example I'd like to use it for curtains that do get pushed aside by dynamic objects, but sometimes these objects get stuck inside the soft body, breaking the gameplay. I would prefer the curtains to be purely decorative, such that they react to the world, but don't influence it.

3) I am quite amazed at the performance of soft-bodies. I expected this to be much slower (I tried with 64*64 vertices in a Debug build and it is FAST). I'm not a physics person, at all, but a couple of years ago I wanted simple cloth / rope simulation, and I couldn't find much on the internet, except one article that showed how to do this with Verlet integration. I adapted that code and it is pretty simple (see https://github.com/ezEngine/ezEngine/blob/dev/Code/Engine/GameEngine/Physics/Implementation/ClothSheetSimulator.cpp). It is also awfully slow because it needs a ton of iterations to be somewhat stable and has a lot of sag. My code was supposed to be the "lightweight" option compared to a full physics simulation, though now I fear it's actually slower :D

3a) Is there something better than Verlet integration, that's more stable and requires fewer iterations? Should it already be and my implementation simply sucks ? Any hints would be appreciated.

3b) For decorative soft-body use cases (flags / curtains) one can often do without collisions, all I want is to set the vertex positions and velocities, do a simulation step and then get the result. I also want full control, for example when a flag is not visible, I just want to pause the simulation entirely. It's even useful to do this inside an editor to pre-simulate (and bake) the state of something. That's how I use my cloth simulation code. Is there a way to use Jolt's soft-body simulation code directly, without a Jolt body, without collisions, just gravity? If so, this would be super useful for placing all sorts of decorative pieces of cloth, ropes, cables and such. Would be even better, if one could add a few custom collision planes, so that they don't clip into walls but otherwise are super light-weight to simulate.

**Comment by jrouwe** [comment, marked as answer]
1. I can investigate this. I think it is possible.
2. I still need to implement contact listeners for soft bodies, I could probably have a mass override there just like the regular contact listener.
3a. I'm using [XPBD](https://matthias-research.github.io/pages/tenMinutePhysics/09-xpbd.pdf) which is similar to Verlet but with a small twist that makes it require less iterations. I'm multi threading the simulation (even within a single soft body), so perhaps that is why it is faster?
3b. You could have a separate physics system perhaps just to simulate a single body? (you do need a body). Alternatively you can put the soft body in an object layer that doesn't collide with anything. If you want custom collision planes I would suggest adding these to the physics system in an object layer that collides only with the soft bodies. If you put these things in a separate broad phase tree then they should be cheap to find for the soft bodies and shouldn't interfere with the rest of the objects. Pausing the simulation can be done by manually deactivating the body.

**Reply by jankrassnigg** [reply]
Thank you! Ten Minute Physics has a couple of videos explaining XPBD that suggest that there is a bit of room of improvement in my code, even without multi-threading and better SIMD. Need to give that a try.

**Comment by jrouwe** [comment]
I've done a bit of work on the 'thickness' of soft bodies. You can find it [here](https://github.com/jrouwe/JoltPhysics/tree/feature/soft_body_vtx_radius). The code requires a bit more cleaning up (contains too much copy paste code), but I think it is almost good to go. Feel free to give it a try.

**Reply by jrouwe** [reply]
It has been merged

**Reply by jankrassnigg** [reply]
Thank you! Sorry I'm currently so busy at work, that I didn't get to have a look yet. Will give it a try as soon as I can.

**Reply by jankrassnigg** [reply]
Finally got around to trying it out and it works really nicely! :)

**Reply by jrouwe** [reply]
Cool!

> Is it possible to have the soft-bodies collide with other (dynamic) objects, but NOT push them?

I was thinking about this today, I'm still planning to make official support for this, but a simple workaround is to give the vertices a very small mass. That would effectively prevent the soft body from affecting the rigid body and as long as all vertices are light it should not affect the way the soft body simulates.

**Reply by jrouwe** [reply]
> Is it possible to have the soft-bodies collide with other (dynamic) objects, but NOT push them?

As mentioned in #936, this is possible now through a SoftBodyContactListener.

**Reply by jankrassnigg** [reply]
Awesome. I'll try to take a look sometime, but no guarantees when.
